### ISSUE TITLE
[CUDA] illegal memory read on batch_norm_backward_kernel

### ISSUE THREAD

**Issue by kokol16** [pytorch/pytorch#189831]
### 🐛 Describe the bug

The CUDA native_batch_norm_backward path launches over the channel dimension even when the reshaped input has zero elements and never checks that save_invstd is nonempty, so the kernel reads it through a null pointer.

```python
import torch

torch.cuda.init()
gpu = torch.device('cuda')

input = torch.full((4, 2, 0, 3), 1.0, device=gpu)
grad_out = torch.zeros(0, device=gpu)
weight = torch.full((2,), float('nan'), device=gpu)
running_mean = torch.full((2,), 1.0, device=gpu)
save_mean = torch.full((2,), 1e25, device=gpu)
save_invstd = torch.zeros(0, device=gpu)
torch.ops.aten.native_batch_norm_backward(
    grad_out, input, weight, running_mean, None,
    save_mean, save_invstd, True, float('-inf'),
    [True, False, False],
)
torch.cuda.synchronize()
```

```
========= COMPUTE-SANITIZER
========= Invalid __global__ read of size 4 bytes
=========     at void at::native::batch_norm_backward_kernel<float, float, float, int>(torch::headeronly::detail::GenericPackedTensorAccessor<torch::headeronly::detail::TensorAccessor<c10::ArrayRef<long>, const T1, (unsigned long)2, torch::headeronly::DefaultPtrTraits, T4>, at::detail::IndexBoundsCheck<(unsigned long)3, T4>, const T1, (unsigned long)3, torch::headeronly::DefaultPtrTraits, T4>, torch::headeronly::detail::GenericPackedTensorAccessor<torch::headeronly::detail::TensorAccessor<c10::ArrayRef<long>, const T1, (unsigned long)2, torch::headeronly::DefaultPtrTraits, T4>, at::detail::IndexBoundsCheck<(unsigned long)3, T4>, const T1, (unsigned long)3, torch::headeronly::DefaultPtrTraits, T4>, torch::headeronly::detail::GenericPackedTensorAccessor<torch::headeronly::detail::TensorAccessor<c10::ArrayRef<long>, T1, (unsigned long)2, torch::headeronly::DefaultPtrTraits, T4>, at::detail::IndexBoundsCheck<(unsigned long)3, T4>, T1, (unsigned long)3, torch::headeronly::DefaultPtrTraits, T4>, torch::headeronly::detail::GenericPackedTensorAccessor<torch::headeronly::detail::TensorAccessor<c10::ArrayRef<long>, T2, (unsigned long)0, torch::headeronly::DefaultPtrTraits, T4>, at::detail::IndexBoundsCheck<(unsigned long)1, T4>, T2, (unsigned long)1, torch::headeronly::DefaultPtrTraits, T4>, torch::headeronly::detail::GenericPackedTensorAccessor<torch::headeronly::detail::TensorAccessor<c10::ArrayRef<long>, T2, (unsigned long)0, torch::headeronly::DefaultPtrTraits, T4>, at::detail::IndexBoundsCheck<(unsigned long)1, T4>, T2, (unsigned long)1, torch::headeronly::DefaultPtrTraits, T4>, torch::headeronly::detail::GenericPackedTensorAccessor<torch::headeronly::detail::TensorAccessor<c10::ArrayRef<long>, const T2, (unsigned long)0, torch::headeronly::DefaultPtrTraits, T4>, at::detail::IndexBoundsCheck<(unsigned long)1, T4>, const T2, (unsigned long)1, torch::headeronly::DefaultPtrTraits, T4>, torch::headeronly::detail::GenericPackedTensorAccessor<torch::headeronly::detail::TensorAccessor<c10::ArrayRef<long>, const T2, (unsigned long)0, torch::headeronly::DefaultPtrTraits, T4>, at::detail::IndexBoundsCheck<(unsigned long)1, T4>, const T2, (unsigned long)1, torch::headeronly::DefaultPtrTraits, T4>, torch::headeronly::detail::GenericPackedTensorAccessor<torch::headeronly::detail::TensorAccessor<c10::ArrayRef<long>, const T2, (unsigned long)0, torch::headeronly::DefaultPtrTraits, T4>, at::detail::IndexBoundsCheck<(unsigned long)1, T4>, const T2, (unsigned long)1, torch::headeronly::DefaultPtrTraits, T4>, torch::headeronly::detail::GenericPackedTensorAccessor<torch::headeronly::detail::TensorAccessor<c10::ArrayRef<long>, const T3, (unsigned long)0, torch::headeronly::DefaultPtrTraits, T4>, at::detail::IndexBoundsCheck<(unsigned long)1, T4>, const T3, (unsigned long)1, torch::headeronly::DefaultPtrTraits, T4>, torch::headeronly::detail::GenericPackedTensorAccessor<torch::headeronly::detail::TensorAccessor<c10::ArrayRef<long>, const T3, (unsigned long)0, torch::headeronly::DefaultPtrTraits, T4>, at::detail::IndexBoundsCheck<(unsigned long)1, T4>, const T3, (unsigned long)1, torch::headeronly::DefaultPtrTraits, T4>, bool, T3)+0x90
=========     by thread (0,1,0) in block (0,0,0)
=========     Address 0x0 is out of bounds
=========     and is 140166624706560 bytes before the nearest allocation at 0x7f7b15e00000 of size 2097152 bytes
```

### Versions

Collecting environment information...

/usr/lib/x86_64-linux-gnu/libcudnn_ops.so.9.8.0
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: False
Caching allocator config: N/A

CPU:
Architecture:                            x86_64
CPU op-mode(s):                          32-bit, 64-bit
Address sizes:                           46 bits physical, 57 bits virtual
Byte Order:                              Little Endian
CPU(s):                                  104
On-line CPU(s) list:                     0-103
Vendor ID:                               GenuineIntel
Model name:                              Intel(R) Xeon(R) Gold 5320 CPU @ 2.20GHz
CPU family:                              6
Model:                                   106
Thread(s) per core:                      2
Core(s) per socket:                      26
Socket(s):                               2
Stepping:                                6
CPU(s) scaling MHz:                      39%
CPU max MHz:                             3400.0000
CPU min MHz:                             800.0000
BogoMIPS:                                4400.00
Flags:                                   fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 invpcid_single ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb intel_pt avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local split_lock_detect wbnoinvd dtherm ida arat pln pts avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg tme avx512_vpopcntdq la57 rdpid fsrm md_clear pconfig flush_l1d arch_capabilities
Virtualization:                          VT-x
L1d cache:                               2.4 MiB (52 instances)
L1i cache:                               1.6 MiB (52 instances)
L2 cache:                                65 MiB (52 instances)
L3 cache:                                78 MiB (2 instances)
NUMA node(s):                            2
NUMA node0 CPU(s):                       0-25,52-77
NUMA node1 CPU(s):                       26-51,78-103
Vulnerability Gather data sampling:      Mitigation; Microcode
Vulnerability Indirect target selection: Mitigation; Aligned branch/return thunks
Vulnerability Itlb multihit:             Not affected
Vulnerability L1tf:                      Not affected
Vulnerability Mds:                       Not affected
Vulnerability Meltdown:                  Not affected
Vulnerability Mmio stale data:           Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Reg file data sampling:    Not affected
Vulnerability Retbleed:                  Not affected
Vulnerability Spec rstack overflow:      Not affected
Vulnerability Spec store bypass:         Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:                Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:                Mitigation; Enhanced / Automatic IBRS; IBPB conditional; PBRSB-eIBRS SW sequence; BHI SW loop, KVM SW loop
Vulnerability Srbds:                     Not affected
Vulnerability Tsa:                       Not affected
Vulnerability Tsx async abort:           Not affected
Vulnerability Vmscape:                   Not affected

Versions of relevant libraries:
[pip3] numpy==2.4.4
[pip3] nvidia-cublas-cu12==12.6.4.1
[pip3] nvidia-cuda-cupti-cu12==12.6.80
[pip3] nvidia-cuda-nvrtc-cu12==12.6.85
[pip3] nvidia-cuda-runtime-cu12==12.6.77
[pip3] nvidia-cudnn-cu12==9.10.2.21
[pip3] nvidia-cufft-cu12==11.3.0.4
[pip3] nvidia-curand-cu12==10.3.7.77
[pip3] nvidia-cusolver-cu12==11.7.1.2
[pip3] nvidia-cusparse-cu12==12.5.4.2
[pip3] nvidia-cusparselt-cu12==0.7.1
[pip3] nvidia-nccl-cu12==2.29.3
[pip3] nvidia-nvjitlink-cu12==12.6.85
[pip3] nvidia-nvtx-cu12==12.6.77
[pip3] torch==2.13.0+cu126
[pip3] triton==3.7.1
[conda] Could not collect

cc @ptrblck @msaroufim @eqy @tinglvv @nWEIdia

**Comment by MDSALMANSHAMS** [comment]
Here `input` is `(4, 2, 0, 3)` (zero-element) while `grad_out`/`save_invstd` are size-0 and `weight` is size 2 — the shapes are mutually inconsistent, and `native_batch_norm_backward` launches the reduction kernel without reconciling them, reading past the size-0 buffers. A shape/size-consistency `TORCH_CHECK` (plus a short-circuit for zero-element input) before the kernel launch would fix the OOB. Would maintainers welcome a PR, and could this be marked `actionable`?

**Comment by Arindamsamanta2004** [comment]
I can reproduce this on the latest main. I'll put up a fix.

Root cause: 
batch_norm_backward_cuda (the CUDA dispatch registered in native_functions.yaml for native_batch_norm_backward) has no zero-element guard. When the input has a zero-size non-channel dimension (e.g. (4, 2, 0, 3)), the fused backward kernel is launched with dim3 blocks(input.size(1)) = 2 blocks, and unconditionally reads save_invstd[plane] -- but save_invstd is an empty tensor with a null data pointer, so that's a null dereference.

There's already a zero-element guard in _batch_norm_impl_index_backward, but calling torch.ops.aten.native_batch_norm_backward directly bypasses it and hits the unguarded CUDA path.

Fix: Adding an early return in batch_norm_backward_cuda  when input.numel() == 0, returning correctly shaped zero-filled gradients. This mirrors the existing composite-level guard and covers all downstream paths (fused, non-fused, contiguous, channels-last).

@eqy @syed-ahmed @Aidyn-A Does this approach look reasonable to you, or would you prefer handling it differently?

**Comment by Claudio999222** [comment]
Hi, I’d like to work on this issue.

My understanding is that the fused CUDA backward path launches one block per
channel even when the reshaped input has zero feature elements. In training
mode, batch_norm_backward_kernel then unconditionally reads
save_mean[plane] and save_invstd[plane], while an empty saved-stat tensor is
represented by a null packed accessor.

My initial plan is to:

1. add host-side validation before launching the fused CUDA kernel.
2. ensure the saved mean and inverse standard deviation contain one value per
   input channel in training mode.
3. add a CUDA regression test verifying that invalid saved statistics produce
   a RuntimeError rather than an illegal device memory access.

Would strict saved-stat shape validation be the preferred behavior, or should
the empty-input case return empty/zero gradients instead? :)

**Comment by malfet** [comment]
Duplicate of #195547. This is one instance of a class of missing shape/size checks that lead to out-of-bounds tensor access — here surfaced by compute-sanitizer on a CUDA kernel rather than by ASAN on CPU, but the same missing bounds contract. The class is now tracked under the umbrella issue #195547, which also explains why per-operator TORCH_CHECK PRs are not the fix being sought. Please follow #195547.
